### PR TITLE
Add user deletion to DBAM

### PR DESCRIPTION
**Pull request by janboll**
This enables the deletion of provisioned users.
Content created by these users will not be deleted, but ownership transferred to the admin user. Schemas are retained as well.

### CODE REVIEW

File: reconcile/database_access_manager.py [hunk 1]
```diff
@@ -77,6 +77,9 @@ def _get_db(self) -> str:
     def _get_user(self) -> str:
         return self.db_access.username
 
+    def _get_admin_user(self) -> str:
+        return self.admin_connection_parameter.user
```
Comment (hemslo) [inline reconcile/database_access_manager.py:81]: readonly property is more common:

```python
    @property
    def admin_user(self) -> str:
        return self.admin_connection_parameter.user
```

File: reconcile/database_access_manager.py [hunk 2]
```diff
@@ -105,8 +122,13 @@ def _generate_db_access(self) -> str:
...
     def generate_script(self) -> str:
-        x = self._generate_create_user() + "\n" + self._generate_db_access()
-        return x
+        if not self.db_access.delete:
+            script = self._generate_create_user() + "\n" + self._generate_db_access()
+        else:
+            script = (
+                self._generate_revoke_db_access() + "\n" + self._generate_delete_user()
+            )
+        return script
```
Comment (hemslo) [inline reconcile/database_access_manager.py:131]: Now this method has 2 responsibilities `create` and `delete`, maybe make it explicit by extracting a provision script and a deprovision script like:

```python
    def provision_script(self) -> str:
        return self._generate_create_user() + "\n" + self._generate_db_access()

    def deprovision_script(self) -> str:
        return self._generate_revoke_db_access() + "\n" + self._generate_delete_user()

    def generate_script(self) -> str:
        return (
            self.deprovision_script()
            if self.db_access.delete
            else self.provision_script()
        )
```

File: reconcile/database_access_manager.py (hunk 1 above)
Comment (janboll) [reply reconcile/database_access_manager.py:81]: Hm, these are not properties of `PSQLScriptGenerator`, but helpers to make accessing related data easier.

File: reconcile/database_access_manager.py [hunk 3]
```diff
@@ -94,9 +97,23 @@ def _generate_create_user(self) -> str:
...
+\\c "{self._get_db()}"
+REASSIGN OWNED BY "{self._get_user()}" TO "{self._get_admin_user()}";
+DROP ROLE IF EXISTS "{self._get_user()}";\\gexec"""
+
+    def _generate_revoke_db_access(self) -> str:
+        statements: list[str] = ["\n"]
+        for access in self.db_access.access or []:
+            statement = f'REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA "{access.target.dbschema}" FROM "{self._get_user()}";\n'
+            statements.append(statement)
+        return "".join(statements)
```
[the remark this answers is not in the document]
Comment (janboll) [reply reconcile/database_access_manager.py:115]: :+1:

File: reconcile/database_access_manager.py (hunk 1 above)
Comment (hemslo) [reply reconcile/database_access_manager.py:81]: It's rare in python to define getter functions, https://stackoverflow.com/questions/2627002/whats-the-pythonic-way-to-use-getters-and-setters.

Such internal helpers can be like this

```python
@property
def _admin_user(self) -> str:
    return self.admin_connection_parameter.user
```

File: reconcile/database_access_manager.py (hunk 2 above)
Comment (janboll) [reply reconcile/database_access_manager.py:131]: :+1:

**Pull request merged**